# E-Commerce Logs Analysis

## Project Objective

The objective of this project is to analyze the website logs from an e-commerce store to obtain useful insight into customer behaviour.

## Data Source

The dataset was obtained here: https://www.kaggle.com/datasets/kzmontage/e-commerce-website-logs and the data has an Open Database license: https://opendatacommons.org/licenses/odbl/1-0/

In [1]:
-- Specifying database
USE EcommerceLogsDb

Commands completed successfully.

Total execution time: 00:00:00.022

In [2]:
-- preview of table
SELECT TOP 3*
FROM logs;

(3 rows affected)

Total execution time: 00:00:00.388

access_date duration_seconds Proto IP Src_IP_type Src_Pt Bytes Accessed_From country membership Languages Sales Returned Returned_Amount 2016-11-01 09:58:00.0000000 2533 TCP 1.10.195.126 EXT_SERVER 8082 20100 Chrome Albania Normal Albanian 261.96 No 0 2016-11-01 09:59:00.0000000 4034 TCP 1.1.217.211 OPENSTACK_NET 56978 20500 Mozilla Firefox Albania Normal Albanian 731.94 No 0 2016-11-01 09:59:00.0000000 1525 TCP 1.115.198.107 EXT_SERVER 8082 90100 Mozilla Firefox Albania Normal Albanian 14.62 No 0

In [3]:
-- Which country had the most sales (transactions) and sales ($ amount)?
-- showing country rankings for both metrics using window functions
SELECT country, 
    COUNT(*) AS sales_transaction, 
    DENSE_RANK() OVER(ORDER BY COUNT(*) DESC) AS sales_transaction_rank, 
    SUM(Sales) AS sales_$, 
    DENSE_RANK() OVER(ORDER BY SUM(Sales) DESC) AS sales_$_rank
FROM logs
GROUP BY country
ORDER BY sales_$ DESC;

(102 rows affected)

Total execution time: 00:00:00.319

country sales_transaction sales_transaction_rank sales_$ sales_$_rank Austria 2674 1 1296463.3414999906 1 Belgium 2604 2 1154321.3705999977 2 Argentina 2603 3 1140238.6927999987 3 Australia 2520 4 1139576.1649999968 4 United States 2231 12 1132496.3800000001 5 Brazil 2604 2 1129919.2000999975 6 Norway 2160 13 1129793.889999999 7 Netherlands 2292 8 1106623.810000002 8 Spain 2232 11 1102578.460000004 9 Puerto Rico 2232 11 1089656.3800000006 10 Canada 2436 6 1078310.4730000005 11 United Kingdom 2232 11 1069649.1000000013 12 Bulgaria 2520 4 1068255.9588000018 13 Republic of Korea 2232 11 1063598.8800000008 14 Singapore 2232 11 1062442.5199999993 15 Suriname 2016 18 1037655.410000003 16 Mexico 2232 11 1034845.1700000012 17 Turkmenistan 2088 16 1032696.5800000039 18 Belize 2352 7 1009231.1078000019 19 Sweden 2148 14 1006956.3000000026 20 Chile 2604 2 998301.8032000033 21 Russian Federation 1944 20 989007.5500000018 22 Thailand 2004 19 984841.0600000018 23 Portugal 1944 20 955279.5800000016 24 Saint Lucia 2016 18 953605.8500000038 25 Mauritius 2292 8 951543.6300000029 26 Trinidad and Tobago 1944 20 940220.2900000017 27 Antigua and Barbuda 2268 9 937031.6729000005 28 Romania 2004 19 933676.0400000019 29 Armenia 2086 17 918236.707500002 30 Barbados 2099 15 914011.7824000027 31 Uruguay 2016 18 898636.5600000003 32 New Zealand 2088 16 897640.3300000007 33 Paraguay 1944 20 887757.4500000015 34 Luxembourg 2232 11 869798.3930000014 35 Ukraine 2016 18 860557.6200000015 36 Bahamas 1932 21 858137.5198000026 37 Malta 2232 11 849198.6836000021 38 Panama 1800 25 827407.4100000006 39 Poland 1728 29 823874.8300000004 40 Colombia 2514 5 823458.3767000027 41 Albania 1848 24 816714.1692000002 42 Guatemala 2160 13 811037.0116000031 43 Bahrain 1764 27 800315.182500001 44 Belarus 1763 28 793122.7560000019 45 Saint Vincent and Grenadines 1800 25 782619.9400000017 46 Greece 2232 11 781112.8694000007 47 Slovenia 1511 33 779910.2300000001 48 Slovakia 1584 30 774529.1199999993 49 Switzerland 1512 32 711132.8300000015 50 Ecuador 2232 11 704434.9145000005 51 Uzbekistan 1584 30 700289.7400000006 52 South Africa 1440 34 699497.384000001 53 Japan 2233 10 692880.2690000008 54 Israel 2232 11 678394.283000001 55 Grenada 1860 23 677864.1923000009 56 Iceland 2292 8 673693.6802000001 57 Guyana 1799 26 673308.8582000005 58 Kuwait 1800 25 668267.5795000002 59 Kyrgyzstan 1872 22 666682.2039000005 60 Ireland 2160 13 649537.6800000011 61 France 2160 13 643784.2059999998 62 Kazakhstan 1872 22 641065.9040000006 63 Hungary 1860 23 634635.7702 64 Costa Rica 2160 13 632901.6622000004 65 Finland 2088 16 630622.2634000012 66 Italy 2232 11 625664.3300000008 67 Seychelles 1296 36 620275.4700000003 68 Azerbaijan 1344 35 610523.8879999998 69 Germany 1872 22 601330.1426000008 70 El Salvador 1728 29 593740.2684 71 Lithuania 1572 31 572731.2999999988 72 Serbia 1296 36 564452.1499999986 73 Qatar 1068 40 554514.8799999999 74 Latvia 1511 33 549710.9939999995 75 Aruba 1176 38 537197.9709999991 76 Philippines 1079 39 522248.86000000004 77 Czech Republic 1932 21 517375.4399999993 78 Cuba 1728 29 500943.9629999991 79 Georgia 1584 30 499963.55559999857 80 Denmark 1584 30 449218.213999999 81 Estonia 1512 32 424996.7387999989 82 Croatia 1572 31 416275.25939999963 83 Sri Lanka 792 42 395918.2999999992 84 Montenegro 720 43 357511.14999999985 85 Jamaica 1224 37 347578.4049999999 86 Kiribati 792 42 323550.16689999966 87 Cyprus 1067 41 281314.48579999973 88 Turkey 504 44 280994.1200000001 89 Nicaragua 432 45 223247.22000000006 90 Maldives 720 43 217764.10799999998 91 United Arab Emirates 432 45 186636.71000000022 92 Fiji 792 42 186410.44499999998 93 Oman 216 46 105546.17000000007 94 San Marino 216 46 98001.99000000008 95 Saint Kitts and Nevis 216 46 87428.36000000006 96 Bosnia and Herzegovina 168 47 60497.63000000005 97 Cabo Verde 84 48 36562.07199999999 98 Macau 72 49 28242.788 99 Mongolia 60 50 22035.249999999985 100 Dominica 72 49 18743.108999999993 101 Russia 8 51 1264.5840000000003 102

From the table above, the top 4 countries for sales\_transaction and sales\_$ are Austria, Belgium, Argentina and Australia. The number of sales transactions seems highly correlated with sales ($) with some exceptions like the United States which is ranked 12th in number of transactions and 5th in sales ($).

Let's look at the average sale amount per country and compare to the worldwide average. The difference between the sale amount for each country and the worldwide average will also be computed.

In [4]:
-- average sales ($) per country
SELECT  
    country, AVG(Sales) AS avg_sales_$,
    -- subquery in SELECT to determine worldwide avg sales ($)
    (SELECT AVG(Sales) FROM logs) AS world_avg_sales_$,
    -- difference bewteen country avg and worldwide avg with another SELECT subquery
    (AVG(Sales) - (SELECT AVG(Sales) FROM logs)) AS difference
FROM logs
GROUP BY country
ORDER BY difference DESC;

(102 rows affected)

Total execution time: 00:00:00.380

country avg_sales_$ world_avg_sales_$ difference Turkey 557.5280158730159 411.3464487919273 146.18156708103123 Norway 523.052726851852 411.3464487919273 111.70627805986732 Qatar 519.2086891385765 411.3464487919273 107.86224034659182 Nicaragua 516.7759722222224 411.3464487919273 105.4295234302378 Slovenia 516.1550165453342 411.3464487919273 104.80856775334956 Suriname 514.7100248015875 411.3464487919273 103.36357600960281 Russian Federation 508.74873971193404 411.3464487919273 97.4022909199494 United States 507.6182787987449 411.3464487919273 96.27183000676024 Sri Lanka 499.89684343434374 411.3464487919273 88.5503946423591 Montenegro 496.5432638888891 411.3464487919273 85.19681509690446 Turkmenistan 494.5864846743292 411.3464487919273 83.24003588234456 Spain 493.9867652329743 411.3464487919273 82.64031644098964 Thailand 491.4376546906189 411.3464487919273 80.09120589863426 Portugal 491.3989609053497 411.3464487919273 80.05251211336508 Slovakia 488.97040404040354 411.3464487919273 77.6239552484189 Oman 488.639675925926 411.3464487919273 77.29322713394134 Puerto Rico 488.19730286738354 411.3464487919273 76.8508540753989 South Africa 485.7620722222221 411.3464487919273 74.41562343023747 Austria 484.8404418474198 411.3464487919273 73.49399305543517 Philippines 484.0119184430032 411.3464487919273 72.66546965101855 Trinidad and Tobago 483.65241255144014 411.3464487919273 72.3059637594555 Netherlands 482.8201614310642 411.3464487919273 71.47371263907957 United Kingdom 479.2334677419351 411.3464487919273 67.88701894995046 Seychelles 478.6076157407402 411.3464487919273 67.26116694875554 Poland 476.7794155092592 411.3464487919273 65.43296671727455 Republic of Korea 476.52279569892477 411.3464487919273 65.17634690694013 Singapore 476.00471326164876 411.3464487919273 64.65826446966412 Saint Lucia 473.01877480158794 411.3464487919273 61.672326009603296 Switzerland 470.32594576719566 411.3464487919273 58.97949697521102 Sweden 468.7878491620112 411.3464487919273 57.44140037002654 Romania 465.90620758482976 411.3464487919273 54.559758792845116 Mexico 463.6403091397844 411.3464487919273 52.29386034779975 Panama 459.670783333333 411.3464487919273 48.32433454134838 Aruba 456.80099574829944 411.3464487919273 45.4545469563148 Paraguay 456.6653549382711 411.3464487919273 45.31890614628645 Azerbaijan 454.25884523809526 411.3464487919273 42.91239644611062 San Marino 453.7129166666667 411.3464487919273 42.36646787468203 Bahrain 453.6934141156462 411.3464487919273 42.34696532366155 Australia 452.21276388888873 411.3464487919273 40.86631509690409 Belarus 449.87110380034017 411.3464487919273 38.52465500835552 Uruguay 445.75226190476235 411.3464487919273 34.40581311277771 Bahamas 444.17055890269114 411.3464487919273 32.8241101107065 Belgium 443.28777672811026 411.3464487919273 31.941327936125617 Canada 442.65618760262726 411.3464487919273 31.309738810642614 Uzbekistan 442.10210858585873 411.3464487919273 30.755659793874088 Albania 441.9448967532468 411.3464487919273 30.59844796126214 Armenia 440.19017617449674 411.3464487919273 28.8437273825121 Argentina 438.04790349596567 411.3464487919273 26.701454703981028 Serbia 435.5340663580239 411.3464487919273 24.18761756603925 Barbados 435.4510635540732 411.3464487919273 24.104614762088545 Cabo Verde 435.2627619047618 411.3464487919273 23.916313112777175 Saint Vincent and Grenadines 434.78885555555553 411.3464487919273 23.442406763570887 Brazil 433.91674350998466 411.3464487919273 22.570294718000014 United Arab Emirates 432.02942129629633 411.3464487919273 20.682972504311692 New Zealand 429.9043726053637 411.3464487919273 18.557923813379034 Belize 429.09485875850316 411.3464487919273 17.748409966518523 Ukraine 426.86389880952385 411.3464487919273 15.517450017539204 Bulgaria 423.9110947619049 411.3464487919273 12.564645969920264 Mauritius 415.15865183246063 411.3464487919273 3.8122030404759926 Antigua and Barbuda 413.1532949294536 411.3464487919273 1.8068461374689377 Kiribati 408.5229380050506 411.3464487919273 -2.

Turkey has the largest positive difference in average $ per sale with $146/sale greater than the worldwide average. In contrast, Russia has the largest negative difference with $253/sale less than the world average.

We'll now look at returns of purchased items for each country.

In [14]:
-- CTE for value of sales returned
WITH returned_$ AS (
    SELECT country,
        -- Sum of total dollars returned
        SUM(Returned_Amount) AS returned_amount_$,
        -- Percentage of dollars returned over dollars sold
        -- Sales column doesn't include amount for returned items
        SUM(Returned_Amount) / (SUM(Sales) + SUM(Returned_Amount)) AS pct_$_returned
    From logs
    GROUP BY country),

-- CTE for number of sales returned
returned_orders AS (
    SELECT t1.country, 
        -- counting all rows (number of transactions)
        COUNT(*) AS sales_transaction,
        -- counting number of order returns 
        COUNT(t2.Returned) AS orders_returned,
        -- calculating percentage of orders that were returned
        1.0 * COUNT(t2.Returned) / COUNT(*) AS pct_orders_returned
    FROM logs as t1
    LEFT JOIN (SELECT duration_seconds, ip, country, Returned
    FROM logs 
    WHERE Returned = 'Yes') AS t2
    ON t1.duration_seconds = t2.duration_seconds
    AND t1.ip = t2.ip
    GROUP BY t1.country)

-- final query with percent of orders returned and percent of sales refunded
SELECT TOP 5 CTE1.country, pct_orders_returned, pct_$_returned 
FROM returned_$ AS CTE1
    INNER JOIN returned_orders AS CTE2
    ON CTE1.country = CTE2.country 
ORDER BY pct_orders_returned DESC, pct_$_returned DESC;



Warning: Null value is eliminated by an aggregate or other SET operation.

(5 rows affected)

Total execution time: 00:00:00.507

country pct_orders_returned pct_$_returned Nicaragua 0.212962962962 0.21179128180131437 Oman 0.212962962962 0.15195254761361243 Montenegro 0.190277777777 0.2197341784693777 Puerto Rico 0.184139784946 0.15605512400495508 Poland 0.182870370370 0.14169463345436817

Nicaragua and Oman have the highest percentage of orders returned at 21.3%. This represents a possible opportunity to increase revenue by making the return policy more strict to attempt to lower the number of returned orders. The affect that a stricter return policy would have on customer experience should also be taken into account though.

We will next look at the number of returns for each language. This will help inform us of any language/communication issues thay may be causing a higher rate of purchase returns.

In [28]:
SELECT TOP 5 t1.Languages, 
        -- calculating percentage of orders that were returned
        ROUND((1.0 * COUNT(t2.Returned) / COUNT(*)), 2) AS pct_orders_returned
FROM logs as t1
LEFT JOIN (SELECT duration_seconds, ip, country, Returned
FROM logs 
WHERE Returned = 'Yes') AS t2
ON t1.duration_seconds = t2.duration_seconds
AND t1.ip = t2.ip
GROUP BY t1.Languages
ORDER BY pct_orders_returned DESC;

Warning: Null value is eliminated by an aggregate or other SET operation.

(5 rows affected)

Total execution time: 00:00:00.122

Languages pct_orders_returned marathi 0.220000000000 kinyarwanda 0.180000000000 romanian 0.180000000000 Slovak 0.170000000000 polish 0.170000000000

Marathi, a language spoken in India, has the highest percentage for orders returned (4% above the second language). A statistical test should be performed to determine if this percentage is statistically significantly greater than the 2nd highest percentage. If it is, this indicates that resources should be spend on improving the website for that language to lower rate of sale returns.